text/0000-road-to-typescript.md

```diff
@@ -0,0 +1,128 @@
+
+## Motivation
+
+> Why are we doing this? What use cases does it support? What is the expected
+outcome?
+
+## Detailed design
+
+TODO: give this an intro
+
+### Constraints
+
+The following are our hard constraints in supporting TypeScript with Ember:
+
+- We must not compromise our commitment to “stability without stagnation” and our strong Semantic Versioning guarantees. 
+- Using TypeScript should never be mandatory for anyone who uses Ember.
+- TypeScript support may never *degrade* the experience of JavaScript users, and wherever possible it should benefit both JavaScript and TypeScript developers.
+
+### Roadmap
+
+With these constraints in mind, the path to full TypeScript support in Ember requires RFCs to address each of the following points:
+
+- **What “official support” entails.** Does it require that all officially published libraries ship TypeScript types? Does it require a specific support policy for all officially published libraries? Should the RFC template be updated in any way to account for TypeScript API constraints? etc.
+
+- **The set of supported types.** Specifically, should the types support the whole API surface of Ember including both Classic and Octane features, or only Octane features? For example: mixins are not part of the Octane programming model *and* are a major source of the complexity of Ember’s current types; should they be included at all in types shipped with 
+
+- **Migration from DefinitelyTyped.** TypeScript users in the Ember community currently rely on the types in DefinitelyTyped. We need to define a transition story for them, which is closely related to the previous point. If we drop support for features currently supported on DefinitelyTyped, or make even well-motivated changes, how will users migrate successfully to them?
+
+- **Semantic Versioning of types and TypeScript compiler version support.** There is currently no widely-used definition for semantic versioning of TypeScript types. Additionally, we must define a support policy for TypeScript versions. (See also further discussion [below](#semantic-versioning-and-supported-typescript-versions).)
+
+- **Supported compilation modes.** What should the out-of-the-box settings be for compiling Ember apps and addons written in TypeScript? This will need to be written with an eye to Embroider and the v2 Package Format as end goals, but without assuming them as available at the starting point. It must also address the transition path from current ember-cli-typescript (if any). It *must* address interoperation hazards between Babel and TypeScript (for example, in the decorators implementations).
```

Comment:
This should also call out `isolatedModules` as an obvious thing we need to address.